backend/LeetWars.Core/LeetWars.Core.BLL/Services/ChallengeService.cs

```diff
@@ -40,27 +56,85 @@ public async Task<ICollection<ChallengePreviewDto>> GetChallengesAsync(Challenge
+                    ChallengesProgress.Completed => challenges.Where(challenge => challenge.Versions.Any(version =>
+                        version.Solutions.Any(solution =>
+                            solution.User != null && solution.User.Uid == userId && solution.SubmittedAt.HasValue &&
+                            solution.SubmittedAt.Value != DateTime.MinValue))),
                     _ => challenges
                 };
             }
 
+            if (page is not null && page.PageSize > 0 && page.PageNumber > 0)
+            {
+                challenges = challenges.Skip(page.PageSize * (page.PageNumber - 1))
+                    .Take(page.PageSize);
+            }
 
             var filteredChallenges = await challenges.ToListAsync();
+
+            //filter runs on the client because filters.TagsIds[] didn't pass to the server and SQL query get error
+            if (filters.TagsIds != null)
+            {
+                filteredChallenges = filteredChallenges.Where(challenge =>
+                    filters.TagsIds.All(
+                        filterTagId => challenge.Tags.Any(tag => tag.Id.Equals(filterTagId))
+                    )
+                ).ToList();
+            }
+
             return _mapper.Map<List<ChallengePreviewDto>>(filteredChallenges);
         }
+
+        public async Task<ChallengePreviewDto> GetChallengeSuggestionAsync(SuggestionSettingsDto settings)
+        {
+            var userId = _userIdGetter.CurrentUserId;
+
+            var challenges = _context.Challenges
+                .Include(challenge => challenge.Tags)
+                .Include(challenge => challenge.Author)
+                .Include(challenge => challenge.Level)
+                .Include(challenge => challenge.Versions)
+                    .ThenInclude(version => version.Language)
+                .Include(challenge => challenge.Versions)
+                    .ThenInclude(version => version.Solutions)
+                        .ThenInclude(solution => solution.User)
+                .Where(c => c.Versions.Any(v => v.LanguageId == settings.LanguageId))
+                .AsQueryable();
+
+            challenges = settings.SuggestionType switch
+            {
+                SuggestionType.Beta => challenges.Where(challenge =>
+                    challenge.Versions.Any(version => version.Status == ChallengeStatus.Beta)),
+                SuggestionType.Fundamentals => challenges.Where(challenge =>
+                    challenge.Level != null && challenge.Level.Name.ToLower().Contains("easy")),
+                SuggestionType.RankUp => challenges.Where(challenge =>
+                    challenge.Level != null && challenge.Level.Name.ToLower().Contains("medium")),
+                SuggestionType.PracticeAndRepeat => challenges.Where(challenge => challenge.Versions.Any(version =>
+                    version.Solutions.Any(solution => solution.User != null && solution.User.Uid == userId))),
+                _ => challenges
```

Comment:
it would be better o change Level Entity to enum? or do we have some other properties in this entity (except Name and Id)?

Reply:
It contains field **reward** for calculating user raiting (likewise kyu in codewars)